### ISSUE TITLE
No batching support for FIFO queues?

### ISSUE THREAD

**Issue by undergroundwebdesigns** [phstc/shoryuken#548]
Hello,

I'm evaluating Shuryuken for use in a high-volume web-app, and I'm concerned that it looks like Shuryuken doesn't support batch fetching for FIFO queues. Specifically [fetcher.rb#61](https://github.com/phstc/shoryuken/blob/master/lib/shoryuken/fetcher.rb#L61) seems to over-ride the batch size (max_number_of_messages) to 1 for any FIFO queue, regardless of whether batch processing is enabled.

The comment explains that this is to ensure FIFO semantics are respected, and while that makes sense to me if the worker is _not_ using batching (where the individual messages returned will be fanned out to multiple threads) if the worker _is_ using batching then all the messages fetched _should_ be sent to the same worker thread, and thus the worker should be able to process them in order. 

Am I misunderstanding something, or is this indeed a limitation in Shuryuken? If so, would you be open to a PR to fix this behaviour?

:pushpin: #563

**Comment by phstc** [comment]
Hi @undergroundwebdesigns 

You raised a good. It isn't much flexible for batch + FIFO. But we can definitely change that.

At that point, we don't know if the messages that are going to come in are for a batch worker or not.

What we could do, is to have an option shoryuken.yml to change the default size for FIFO.

We may need to make that option per queue though. Some people may want to do batch in one FIFO queue, but not on another.

WDYT?

**Comment by undergroundwebdesigns** [comment]
I agree that if a new option were added it would need to be configured per queue / worker. 

It's a shame this can't be tied to whether the worker is consuming in batches or not, as I think it most cases if you're having a worker batch-consume from a FIFO queue you're going to want to set this option, and it would be a little confusing to have to both set that the worker should process in batches _and_ set a separate option for the FIFO batch size, especially since for non-FIFO queues you just set the worker to batch consume and you're done. I can see lots of people getting confused, just setting their worker to consume in batches, and then wonder about the low throughput / high SQS bills that result.

**Comment by phstc** [comment]
@undergroundwebdesigns I agree, the challenge is that the out-of-box implementation "needs" to be flexible. For example, you can have multiple workers per queue, Shoryuken can only determine which worker is going to handle a message after the message is received. Making it hard to decide how to fetch based on workers.


What I've been doing over time, is to make it more  extendable, for example, supporting custom polling strategies.  So users can tailor Shoryuken for their specific needs.

I will be happy to guide you through if you are interesting on making a pull request.

**Comment by undergroundwebdesigns** [comment]
I'm curious in what case having "multiple workers per queue" makes sense, doesn't that just mean that one worker will get some subset of the messages, and another worker will get some other subset of the messages, with no way to determine which message go to which worker? How does Shoryuken determine which worker is going to handle a given message in that case?

**Comment by phstc** [comment]
> doesn't that just mean that one worker will get some subset of the messages, and another worker will get some other subset of the messages, 

correct, it means that ☝️ 

> How does Shoryuken determine which worker is going to handle a given message in that case?

The [default worker registry](https://github.com/phstc/shoryuken/blob/master/lib/shoryuken/default_worker_registry.rb#L16-L27) (you can customize this) first tries to find the message respective worker based on the message header. If no header, it tries to find the worker based on the queue name (fallback).

When using headers, you can use multiple workers per queue. `WorkerA.perform_async` enqueues a message with a header telling that it should to be consumed by `WorkerA`. 

Multiple workers per queue is very common when using Active Job.

Does that make sense?

But I have good news, I always just going through the code and I found out Shoryuken does not support multiple workers per queue when using batch, [see this](https://github.com/phstc/shoryuken/blob/master/lib/shoryuken/default_worker_registry.rb#L35). With that said, the fetcher could try to find the worker based on the queue name, then if batch, look for more than one message. I would still make this configure though. 

WDYT?

**Comment by undergroundwebdesigns** [comment]
Ok, that make sense. I wasn't aware that Shoryuken passed a header specifying which worker should process the job, that's the missing piece. Thanks for explaining :)

I think your proposed solution (having the fetcher try to find the worker based on the queue name, and if it's a batch worker pull more than one message even for FIFO queues) would be a great improvement. 👍

**Comment by phstc** [comment]
@undergroundwebdesigns I can't give you an ETA when I will be able to work on that. But I will be happy to guide you through if you are interested on creating a pull request.

**Comment by undergroundwebdesigns** [comment]
Ok cool. We're currently evaluating Shoryuken (and SQS in general), if we decide to go with it we'll almost certainly need this feature, so if that's the case I'll touch base about how best to go about opening that PR.

**Comment by phstc** [comment]
@undergroundwebdesigns I'm very happy with SQS. But there are definitely some gotchas depending on your use case.

If you are also considering Sidekiq, [I have a post comparing Redis vs SQS](http://www.pablocantero.com/blog/2015/03/14/sidekiq-redis-vs-shoryuken-aws-sqs/) that may help you.

What's not updated on that post is the Sidekiq support, Mike Perham does a great job with that.

**Comment by Fito** [comment]
@phstc I added this comment to the PR where the `max_number_of_messages` limitation for FIFO queues was introduced but I'm adding it here as well since the issue seems relevant.
My understanding is that SQS handles hiding messages from a given group ID to ensure messages are processed one by one. I don't think Shoryuken consumers should be limited to pulling one message at a time when working with FIFO queues.
I believe this is a feature of SQS FIFO queues and their API should already be taking care of it.
Thoughts?

**Comment by phstc** [comment]
@Fito I left a comment [here](https://github.com/phstc/shoryuken/blob/34f8ae64bbec53d7fa388540923eff938a66ff5f/lib/shoryuken/fetcher.rb#L52-L60) with some more details.

If that ☝️  is no longer the case, I will be happy to change it or even make it customizable. 

An easy test with a default Shoryuken setup using `concurrency: 25` is to send 1, 2, 3, ... 100 to the group A, then in the worker `puts` the message body and add some arbitrary `sleep rand(1..5)`. You will see the output out of order.

```ruby
# Gemfile
source 'https://rubygems.org'

gem 'shoryuken'
```

```sh
bundle install 
bundle open shoryuken
# comment out the fifo condition https://github.com/phstc/shoryuken/blob/master/lib/shoryuken/fetcher.rb#L61

bundle exec shoryuken sqs create test.fifo
bundle exec shoryuken -q test.fifo -r ./test_worker.rb -c 25
```

```ruby
# test_worker.rb

class TestWorker
  include Shoryuken::Worker
  
  shoryuken_options queue: 'test.fifo', auto_delte: true

  def perform(sqs_msg, body)
    sleep rand(1..5)
    puts body
  end
end

100.times do |i| 
  body = "#{i} -- #{Time.now}" 
  # Time.now just to prevent SQS discarding messages because of deduplication in case of multiple tests
  TestWorker.perform_async(body, message_group_id: 'group a')
end
```

**Comment by phstc** [comment]
@Fito @undergroundwebdesigns if you guys are still interested, I created a patch to make the number of messages customizable for FIFOs. If not set, it will default to 1.

For changing the number of messages:

```ruby
# https://github.com/phstc/shoryuken/wiki/Receive-Message-options

Shoryuken.sqs_client_receive_message_opts = { 
  max_number_of_messages: 10
}
```

### Patch

```diff
From deaf67825dab97b3315d24166d724813b67f2af6 Mon Sep 17 00:00:00 2001
From: Pablo Cantero <pablohstc@gmail.com>
Date: Wed, 24 Apr 2019 16:43:58 -0400
Subject: [PATCH] Make FIFO number of messages customizable

---
 lib/shoryuken/fetcher.rb | 22 +++++++++++++---------
 1 file changed, 13 insertions(+), 9 deletions(-)

diff --git a/lib/shoryuken/fetcher.rb b/lib/shoryuken/fetcher.rb
index 5dd9b84..b974922 100644
--- a/lib/shoryuken/fetcher.rb
+++ b/lib/shoryuken/fetcher.rb
@@ -14,7 +14,7 @@ module Shoryuken
 
         logger.debug { "Looking for new messages in #{queue}" }
 
-        sqs_msgs = Array(receive_messages(queue, [FETCH_LIMIT, limit].min))
+        sqs_msgs = Array(receive_messages(queue, limit))
 
         logger.debug { "Found #{sqs_msgs.size} messages for #{queue.name}" } unless sqs_msgs.empty?
         logger.debug { "Fetcher for #{queue} completed in #{elapsed(started_at)} ms" }
@@ -49,6 +49,16 @@ module Shoryuken
 
       shoryuken_queue = Shoryuken::Client.queues(queue.name)
 
+      options[:max_number_of_messages]  = max_number_of_messages(shoryuken_queue, limit, options)
+      options[:message_attribute_names] = %w[All]
+      options[:attribute_names]         = %w[All]
+
+      options.merge!(queue.options)
+
+      shoryuken_queue.receive_messages(options)
+    end
+
+    def max_number_of_messages(shoryuken_queue, limit, options)
       # For FIFO queues we want to make sure we process one message per group at the time
       # if we set max_number_of_messages greater than 1,
       # SQS may return more than one message for the same message group
@@ -58,16 +68,10 @@ module Shoryuken
       # > in a strict order relative to the message group
       # > (however, messages that belong to different message groups might be processed out of order).
       # > https://docs.aws.amazon.com/AWSSimpleQueueService/latest/SQSDeveloperGuide/using-messagegroupid-property.html
-      options[:max_number_of_messages]  = shoryuken_queue.fifo? ? 1 : max_number_of_messages(limit, options)
-      options[:message_attribute_names] = %w[All]
-      options[:attribute_names]         = %w[All]
 
-      options.merge!(queue.options)
-
-      shoryuken_queue.receive_messages(options)
-    end
+      max_number_of_messages = options[:max_number_of_messages]
+      max_number_of_messages = 1 if max_number_of_messages.nil? && shoryuken_queue.fifo?
 
-    def max_number_of_messages(limit, options)
       [limit, FETCH_LIMIT, options[:max_number_of_messages]].compact.min
     end
 
-- 
2.19.1

```

WDYT? Feel free to create a PR or ask me to in case you want to use it.

**Comment by skcc321** [comment]
Thank you for a great gem. I can be wrong, but....
Why can't we instantiate SQS client inside threads? If we could do that, it will allow people to use grouped messages without limitation concurrency to 1. Everything will be handled by Aws.
In my opinion, here, the consumer is a Thread, not a process, because it processes the message after all. So it is a bit weird that process is a consumer (has SQS client instance)

**Comment by phstc** [comment]
Hi @skcc321 😄 

> Why can't we instantiate SQS client inside threads? If we could do that, it will allow people to use grouped messages without limitation concurrency to 1

The concurrency set to 1 is to guarantee we process one message at a time. How would a SQS client instance per thread be different/prevent that?

> the consumer is a Thread, not a process, because it processes the message after all. So it is a bit weird that process is a consumer (has SQS client instance)

What consumes messages is thread-based, Shoryuken itself is a process, but receiving and consuming messages are thread-based.

**Comment by phstc** [comment]
A plot twist. I ~may~ have lied:

> At that point, we don't know if the messages that are going to come in are for a batch worker or not.

Actually, [Shoryuken does not support multiple workers per queue combined with batch](https://github.com/phstc/shoryuken/blob/master/lib/shoryuken/default_worker_registry.rb#L37-L39). So we know before receiving messages if it is for a batch worker. 

With that, we could enforce the limit set [here](https://github.com/phstc/shoryuken/blob/master/lib/shoryuken/manager.rb#L77).

@undergroundwebdesigns are still interested on that? I can open a PR.

@skcc321 is batch/FIFO also a problem for you?

**Comment by phstc** [comment]
This pull request #563 should fix the problem with FIFO and batched workers. When batched workers, Shoryuken will try to poll the receive message limit as usual.

**Comment by skcc321** [comment]
@phstc, thank you for quick response.
I can be wrong and you can correct me if so, but as I understand from the code
SQS client (consumer) which is located there https://github.com/phstc/shoryuken/blob/master/lib/shoryuken/fetcher.rb#L50 fetches the messages from Aws sqs and after that, it feeds workers here https://github.com/phstc/shoryuken/blob/master/lib/shoryuken/manager.rb#L73 and each worker process messages in out of order (parallelly) (talk about receiving 10 messages at a time)
so
aws takes care about the ordering of grouped messages with multiple consumers itself.
so, for example, we have a single consumer which receives messages and feeds 10 workers with 10 messages from the same group. Order is broken.
If the consumer was at worker level aws grouped FIFO queue will take care about ordering itself.
I know, probably this is not related to the current topic, but anyway...

**Comment by phstc** [comment]
@skcc321 I believe we would still have the problem that the `max_number_of_messages = 1` tries to solve. For non-batched workers, if you receive 10 messages, there will be 10 workers trying to consume them in parallel (threads), which won't preserve the order received from SQS.

For example, if you send three messages (same group: `order-id-123`):

1. create order
2. create a payment
3. mark as shipped

If Shoryuken receives the three messages at once, Shoryuken will try to process them in parallel (thread workers).

The issue reported here, is because Shoryuken also supports [batch](https://github.com/phstc/shoryuken/wiki/Worker-options#batch), which should receive more than one message at once, even when using FIFO. This is fixed in  #563

**Comment by bitbier** [comment]
@phstc I think that maybe you may be misunderstanding what @skcc321 is trying to say. I too have been trying to figure out how to scale out the FIFO worker and looks like we may need to use something different in order to get a higher throughput unless I find a work around in another issue or inspecting the code.

Anyways, I think to clarify what @skcc321 is trying to say is that if you have the fetcher on the worker themselves.

Say, you are running 25 workers and each worker has it's own SQS client. When you pull from the same FIFO queue using max message 1, AWS will send 1 message for a specific message group in the FIFO queue to a worker that is polling at that point in time. That worker will then proceed to process that message. Another worker also polling at that time will receive another message from a different message group in the FIFO queue and continue processing that so on and so forth. This enables concurrency in FIFO and guarantees that only 1 message per message group is being processed at any one time. All of this being said, this still doesn't support batching. Not sure how you can support batching without routing messages after receiving them to an in-memory queue so that a single worker processes 1 message from a group at time, which is what we are currently experiencing today.

**Comment by phstc** [comment]
Hi @bitbier, thanks for raising this.

The latest version solved the problem that was causing [Shoryuken batching](https://github.com/phstc/shoryuken/wiki/Worker-options#batch) to be limited to one message at a time, but that is limited to the FIFO queue constraints + SDK + threads, etc as you described.

> Say, you are running 25 workers, and each worker has its own SQS client.

Having a worker capable of fetching is something that was discussed before, but that's complicated, since it would drastically increase the number of requests to SQS, since each worker would have to receive only one message at a time. We can't receive more and keep them in memory, their visibility timeout could expire and that can cause other problems.

One change that could be done is to have an SQS client per manager, currently, all managers share the same client. That could be done [here](https://github.com/phstc/shoryuken/blob/7b24e1df6758cbae9f6abfd833cbc79608c5c766/lib/shoryuken/launcher.rb#L73).

With that, you would be able to create multiple [processing groups](https://github.com/phstc/shoryuken/wiki/Processing-Groups) each one with their client, and potentially increasing the throughput. For example, 5 groups, each one with concurrency 5. That would increase the number of requests, but not as much as 1 request per worker. WDYT?

**Comment by bitbier** [comment]
Ah gotcha, yeah I think I was misunderstanding how Shoruyken actually worked. I have a metric on how long it takes to process a message from the FIFO queue looking at the time it was first added to the queue and the time it finished and noticed that the time was increasing linearly when there was a burst of messages. Initially I thought that Shoruyken was blocking per fetched messages, but looking at the code the message is handed off the the executor and fetching starts again if there are available threads to process them. I think the stacked/linear execution time is simply the request I/O time required to make that network call to fetch 1 message at a time sequentially.

So in effect, I think Shoryuken is actually doing what I expect it to and there is rather some issue in my own code that I need to address. Sorry to highjack the comment on this closed ticket. Also thanks for helping to clarify some points.
